### ISSUE TITLE
Package group cannot be reinstalled

### ISSUE THREAD

**Issue by Ricky-Tigg** [rpm-software-management/dnf5#501]
**v.:** 4.15.0 | Hello. Package group cannot be reinstalled.
```
$ dnf -v --installed group list | grep -i '^   virt'
$ dnf -v --hidden group list | grep -i '^   virt'
 Virtualization (virtualization)
$ dnf history | grep -i '@virt'
 29 | -qy install @virtualization | 2023-03-27 10:16 | Install | 13
```
_Note:_ `dnf history` **command output truncated by itself**; investigation to figure out on which side the issue sits is out of scope here. Complete command is `dnf -qy install @virtualization`.
```
# dnf -qy reinstall @virtualization
Error: No packages marked for reinstall.
```

**Comment by j-mracek** [comment]
May I ask you to describe what you want to achieve with `# dnf -qy reinstall @virtualization`? Only reinstalling packages installed by the group. Install not installed packages from the group or something else?

**Comment by Ricky-Tigg** [comment]
I should have thought of that. I have reason to believe that along with removals of packages  i had operated, some of the packages that were initially installed along with `@virtualization` have been removed. Then that would match the option _install non-installed packages from that group_.

**Comment by Ricky-Tigg** [comment]
I should have thought of that. I have reason to believe that along with removals of packages  i had operated, some of the packages that were initially installed along with `@virtualization` have been removed. Then that would match the option _reinstall packages from that group that are still installed and install those from that group that are supposed removed_.

**Comment by j-mracek** [comment]
If you want to install uninstalled packages from a group just run `dnf5 group install <group>`. The requirement for install or reinstall is quite rare therefore the question whether we should implement it.
I am proposing to close the issue as a valid but it only resolves too rare use case.

**Comment by Ricky-Tigg** [comment]
Whatever your proposal, no doubt it is relevant. Yet closing would demonstrate that the error `No packages marked for reinstall` is not meant to be fixed in DNF v.4 nor in DNF v.5 if it exits there too, which would be unusual.

**Comment by Ricky-Tigg** [comment]
Whatever your proposal, no doubt it is relevant. Yet closing as it is now, would demonstrate that the error `No packages marked for reinstall` is not meant to be fixed neither in DNF v.4 nor in DNF v.5 if it exists there too, which would be unusual.

**Comment by pkratoch** [comment]
Hi, I think there is nothing that can be done with this issue. The `reinstall` command is not meant for groups (both in documentation and in `--help`, it explicitly mentions package specs and not specs in general) and there is no `reinstall` subcommand for the `group` command. The error message is now also improved, so it's not as confusing as before:
```
# dnf5 reinstall @c-development
Updating and loading repositories:
Repositories loaded.
Failed to resolve the transaction:
Reinstall action for argument "c-development" is not supported.
```

Therefore, I am closing this issue. If you have any more concerns regarding the reinstall behavior, please reopen it.

**Comment by RokeJulianLockhart** [comment]
> https://github.com/rpm-software-management/dnf5/issues/501#issuecomment-2577629749

@pkratoch, that error message doesn't appear if the user attempts to use `sudo dnf5 group reinstall development-tools`. I realise that it states that there's no `reinstall` subcommand, but that doesn't sufficiently elaborate - I had presumed that the verbiage was merely inconsistent between `dnf5` and `dnf5 group` before seeing this issue.

**Comment by pkratoch** [comment]
Hi, I would think the fact that there is no subcommand `reinstall` for command `group` makes it clear enough that there is no `reinstall` action for a group argument. Can you please elaborate on what exactly is the issue?

**Comment by RokeJulianLockhart** [comment]
> Can you please elaborate on what exactly is the issue?

I don't believe that I can elaborate further, nor that it would be particularly useful in actuality.

Rather, might I know *why* there is no `reinstall` command for a group? I ask because I've read this thread, but see no explanation of rationale.

**Comment by pkratoch** [comment]
> Rather, might I know _why_ there is no `reinstall` command for a group? I ask because I've read this thread, but see no explanation of rationale.

I think the main reason is that, in the past, there was not a clear difference between how `group upgrade` and `group reinstall` should behave and even when the `group upgrade` was changed over time, there was never a strong need for `group reinstall`.

The group actions are not that easy to define because there are two possible things that it can affect - the individual packages contained by the group and the packagelist that the group defines. For example, `group upgrade` upgrades the packages and updates the packagelist with new package items. But it does not remove anything from the package list because then a disabled repository might threaten to just remove everything from a group which is often not what was actually wanted.

If we add `group reinstall`, I would expect it should reinstall all packages in a group, but I am not so sure about what to do with the packagelist. I suppose the most logical behavior would be to "synchronize" it with whatever is currently in the repositories, which would be consistent with sequential `remove` and `install`. I would actually be for introducing the `reinstall` action, it would also make sense for https://github.com/rpm-software-management/dnf5/issues/2103 , but it is very low priority now, as there are not many use cases for it, most things can be achieved with `install` or `upgrade`.
